### ISSUE TITLE
Make it easy to 'inherit' config for filters (listener->vhost->route)

### ISSUE THREAD

**Issue by ravenblackx** [envoyproxy/envoy#25903]
The current model of filter config is we *usually* do a complete override of the parent config when we add a vhost or route config. The reason for this is not an aesthetic or intentional one, but rather that the current implementation makes inheritance-based configuration unwieldy and unperformant. It's also impossible to correctly verify that the merged/inherited config is valid at config-time; we don't have access to the merged config until the filter begins work.

For example, see [FileSystemBufferFilterMergedConfig](https://github.com/envoyproxy/envoy/blob/7974560b04172e21999ecd1103f9960b5d2899d4/source/extensions/filters/http/file_system_buffer/filter_config.h#L60-L102) - a horrifying thing that has to be reconstructed for each filter instance.

The premise of this issue is to make it so that during `createRouteSpecificFilterConfigTyped`, we have access to the parent config protos, and thereby can perform a merge relatively easily, just once, at config-time, including validation.

One possible wrinkle with this idea is in a case where the config looks like
```
listener {
  httpfilter{dynamic_my_filter},
  vhost {
    route {
      httpfilter{my_filter{x:1, y:2}}
    }
  }
}
```
An update to the dynamic config in this context may not automatically re-call `createRouteSpecificFilterConfigTyped` for the potentially-impacted 'child' filters (e.g. a hypothetical change from z:3 to z:4 should update both versions of the config), so it may be necessary to also explicitly do that as part of this change.

**Comment by ravenblackx** [comment]
Some clarifications from a slack thread.
* The goal here is not to simplify *finding* the right config - that will still require looking up the route during `decodeHeaders`. It's to simplify and unify *merging* configs, and make that happen at the right time. The ideal outcome is that the operation in `decodeHeaders` can always be simply to use [resolveMostSpecificPerFilterConfig](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/http/utility.h?L583&subtree=true), and have immediate access to an already fully merged config. (Versus currently, if we want to merge headers, we must use a `traverse` function to look at potentially 3 config objects and discern where inheritance should apply.)
* The ECDS problem mentioned is a real problem for this idea, and will make any solution here complicated. Possible options include walking all 'child' configs of any filter reconfigured by ECDS, reconstructing by default, or having 'child' configs at create-time record a 'subscribed' connection to the 'parent' config that triggers a rebuild if the parent config is updated.
* An example of the validation problem mentioned is, e.g.
  ```
  my_filter_at_listener {disabled: true, some_value_that_must_be_set_when_not_disabled: 1234}
  my_filter_at_route {disabled: false}
  ```
  If we validate these configs' properties independently, the latter config appears to be invalid, even though with inheritance from the parent listener it would not be.
  If we validate only at merge-time in the current system, and the config really *is* invalid, it's too late to reject the config, we're already trying to serve traffic with it (because the merge can first happen at `decodeHeaders`).
  One of the goals of this proposal is that we will be able to correctly reject a config update that is invalid after inheritance is applied, and not reject a config that is *only* valid after inheritance is applied, during config load.

**Comment by ravenblackx** [comment]
(Note: I just discovered [getMergedPerFilterConfig](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/http/utility.h?L654&subtree=true) exists; this simplifies the operation slightly, but doesn't merge with the listener-level config, and doesn't address merging performance issues.)

One possible solution to this issue would be to leave the existing structure alone, but add a helper cache utility so that a config merge operation can easily be performed once (per thread? or readwritelock?) after which a shared_ptr<SomeConfig> could be looked up for a given set of route/vhost/listener filter configs. This would eliminate the repeated merge cost, and reduce the complexity of implementing traversal per filter, similar to getMergedPerFilterConfig but with also merging the base listener filter.

One possible issue with this model is that per-route configs are delivered to filters as references, not shared_ptrs, so there would be no easy way to detect when a per-route config no longer exists and should be removed from the cache. This could potentially lead to a slow memory leak, if per-route configs are changed regularly and per-listener configs are not. To resolve this, the cache config object could be a member on a per-listener config object, to reduce lookup complexity, which would prevent memory leaks. Potentially the cache object could require the per-route config objects to inherit from a MergeableConfig type whose destructor could then remove it from any caches.

i.e.
`MyFilterListenerConfig` contains a `shared_ptr<MergedConfigCache<MyFilterListenerConfig, MyFilterRouteConfig, MyFilterMergedConfig>>`
`MyFilterPerRouteConfig` implements `MergeablePerRouteConfig`
`MergeablePerRouteConfig`, upon being merged by a cache object, captures a lambda to remove itself from the cache, which is called during its destructor.
(note, `MyFilterListenerConfig`, `MyFilterRouteConfig` and `MyFilterMergedConfig` may all be the same type, or 2:1, this would be up to the filter implementer.)

**Comment by toddmgreer** [comment]
[getMergedPerFilterConfig](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/http/utility.h?L654&subtree=true) also doesn't help with the inability to verify merged config at config-time. E.g. there's no way to verify that a given field must be set in either per-filter or per-route config.

**Comment by ravenblackx** [comment]
Good point. My proposed cache-based solution also doesn't help with that, since it still postpones the first merge until the filter tries to use the combination of configs, which is too late to verify config validity. So it looks like the original "do something that's probably very complicated" proposal may be the only way that really achieves all the goals for the issue.
